Source/DafnyCore/Compilers/CSharp/Compiler-Csharp.cs

```diff
@@ -2881,9 +2881,27 @@ protected class ClassWriter : IClassWriter {
           }
 
         case BinaryExpr.ResolvedOpcode.LeftShift:
-          opString = "<<"; truncateResult = true; convertE1_to_int = true; break;
+          if (resultType.NormalizeExpand() is
+              BitvectorType { Width: var width and (32 or 64) }) {
+            staticCallString = $"{DafnyHelpersClass}.Bv{width}ShiftLeft";
+            convertE1_to_int = true;
+            truncateResult = true;
```

Comment:
For 32- and 64-bit values, the result should not need to be truncated. However, the optimization to omit truncation for these (and also for 16- and 8-bit values) better belongs in `EmitBitvectorTruncation` (which need not be part of this PR).

Reply:
I'm not sure how to optimize away the truncation. Could you perhaps file an issue with what you have in mind?

Reply:
I took a look at our generated code and apparently we already do the optimization that I suggested here. In particular, the Dafny code

``` dafny
  var x: bv7 := 63;  //    011 1111
  var y: bv8 := 63;  //   0011 1111
  var z: bv9 := 63;  // 0 0011 1111
  // a 5 left-shift of    011 1111 in a  uint8 gives           1110 0000, which after truncation gives            110 0000
  // a 5 left-shift of   0011 1111 in a  uint8 gives           1110 0000 -- no truncation needed
  // a 5 left-shift of 0 0011 1111 in a uint16 gives 0000 0111 1110 0000, which after truncation gives 0000 0001 1110 0000
  print x << 5, " ", y << 5, " ", z << 5, "\n"; // 96 224 480
```

I noticed that the three shift operations are already emitted by the C# and Java compilers as:

``` c#
  (byte)(    ((byte)(x << 5)) & (byte)0x7F     )
  (byte)(    ((byte)(y << 5))                  )
(ushort)(  ((ushort)(z << 5)) & (ushort)0x1FF  )
```

So, we're good.